src/index/client/oauth2.ts

```diff
@@ -0,0 +1,678 @@
+	 */
+	signingSecret?: string;
+
+	/**
+	 * If true, the signature on the access token response is verified.
+	 */
+	verifySignature?: boolean;
+
+}
+
+/**
+ * Parameters required for the [Refresh request](https://openid.net/specs/openid-connect-core-1_0.html#RefreshingAccessToken).
+ */
+export interface RefreshGrantParams extends RefreshTokenGrantorParams {
+
+	/**
+	 * Defines this grant type.
+	 */
+	grantType: 'refresh_token';
+
+}
+
+export interface RefreshTokenGrantorParams extends TokenGrantorParams {
+
+	/**
+	 * The refresh token.
+	 */
+	refreshToken: string;
+
+}
+
+export interface RevocationOptions {
+
+	/**
+	 * If true, uses a GET rather than POST request to revoke an access token.
+	 */
+	useGet?: boolean;
+
+}
+
+export interface TokenGrantorParams extends HasClientId {
+
+	/**
+	 * The client secret for your application.
+	 */
+	clientSecret?: string;
+
+}
+
+export type GrantParams = AuthCodeGrantParams | JwtGrantParams | RefreshGrantParams;
+
+const DEFAULT_GRANT_OPTIONS = Object.freeze({
+	decodeIdToken: true,
+	parseUserInfo: true,
+	responseFormat: 'application/json',
+	verifySignature: true
+});
+
+const DEFAULT_REVOCATION_OPTIONS = Object.freeze({
+	useGet: false
+});
+
+const DEFAULT_REQUEST_CONFIG = Object.freeze({
+	validateStatus: undefined
+});
+
+/**
+ *  An OAuth 2.0 client that implements the [OAuth 2.0 Authorization Framework](https://tools.ietf.org/html/rfc6749) specification.
+ */
+export class OAuth2Client implements AuthClient {
+
+	/**
+	 * The client name.
+	 */
+	protected clientType: string;
+
+	/**
+	 * [Authorization Endpoint](https://tools.ietf.org/html/rfc6749#section-3.1)
+	 */
+	protected authorizationEndpoint?: string;
+
+	/**
+	 * [Token Endpoint](https://tools.ietf.org/html/rfc6749#section-3.2)
+	 */
+	protected tokenEndpoint?: string;
+
+	/**
+	 * [Revocation Endpoint](https://tools.ietf.org/html/rfc7009#section-2)
+	 */
+	protected revocationEndpoint?: string;
+
+	/**
+	 * The environment parameters used for this client.
+	 */
+	protected readonly env: Readonly<Environment>;
+
+	/**
+	 * Creates a new OAuth2 client with the given environment.
+	 *
+	 * @param env The OAuth2 environment parameters.
+	 */
+	constructor(env: Environment) {
+		this.clientType = 'OAuth2';
+		this.env = Object.freeze(Object.assign({}, env));
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public async init() {
+
+		if (!this.env.authorizationEndpoint) {
+			throw new Error('Missing required string parameter: env[authorizationEndpoint]');
+		}
+
+		if (!this.env.revocationEndpoint) {
+			throw new Error('Missing required string parameter: env[revocationEndpoint]');
+		}
+
+		if (!this.env.tokenEndpoint) {
+			throw new Error('Missing required string parameter: env[tokenEndpoint]');
+		}
+
+		this.authorizationEndpoint = this.env.authorizationEndpoint;
+		this.revocationEndpoint = this.env.revocationEndpoint;
+		this.tokenEndpoint = this.env.tokenEndpoint;
+
+	}
+
+	/**
+	 * @inheritdoc
```

Comment:
These should not be necessary, as it inherits the doc by omission of the comment (http://usejsdoc.org/tags-inheritdoc.html). It seems this was for compatibility purposes and should be redundant for us.

However it does not seem to include parameter lists on either method (At least on hover in vscode). Could you test this out? I've added parameter lists before on overridden methods but they did have different method signatures.

Reply:
Discovered that between v5.12.0 and v5.13.0 of tslint some changes were made to the `completed-docs` rule which meant that even overloaded methods required a doc comment.

This is going to be resolved with some new defaults coming soon, however until then it's going to flag up all missing comments even ones that are auto inherited.

Sources:
https://github.com/palantir/tslint/issues/4573
https://github.com/palantir/tslint/pull/4563

Reply:
Cool - I'm leaving this PR for now.